File: problems/2025/09-03/reverse-a-doubly-linked-list/solution.py
from typing import Optional, List


class Node:
    def __init__(self, data: int):
        self.data = data
        self.next: 'Optional[Node]' = None
        self.prev: 'Optional[Node]' = None


class Solution:
    def reverse(self, head: Optional[Node]) -> Optional[Node]:
        if head is None or head.next is None:
            return head

        curr = head
        temp: Optional[Node] = None

        # Swap prev and next for every node
        while curr is not None:
            temp = curr.prev
            curr.prev = curr.next
            curr.next = temp
            # move to the new prev (originally next)
            curr = curr.prev

        # temp is at the node before new head
        if temp is not None:
            head = temp.prev

        return head


# Helpers for quick checks

def build_dll(values: List[int]) -> Optional[Node]:
    head: Optional[Node] = None
    prev: Optional[Node] = None
    for v in values:
        n = Node(v)
        if head is None:
            head = n
        if prev is not None:
            prev.next = n
            n.prev = prev
        prev = n
    return head


def to_array(head: Optional[Node]) -> List[int]:
    out: List[int] = []
    while head is not None:
        out.append(head.data)
        head = head.next
    return out


if __name__ == "__main__":
    s = Solution()

    h = build_dll([3, 4, 5])
    h = s.reverse(h)
    print(to_array(h))  # [5, 4, 3]

    h = build_dll([75, 122, 59, 196])
    h = s.reverse(h)
    print(to_array(h))  # [196, 59, 122, 75]

File: problems/2025/09-03/reverse-a-doubly-linked-list/solution.js
function Node(data) {
  this.data = data;
  this.next = null;
  this.prev = null;
}

function reverse(head) {
  if (!head || !head.next) return head;

  let curr = head;
  let temp = null;

  while (curr) {
    temp = curr.prev;
    curr.prev = curr.next;
    curr.next = temp;
    curr = curr.prev; // move to new prev (originally next)
  }

  if (temp) head = temp.prev;
  return head;
}

// Helpers
function buildDLL(arr) {
  let head = null, prev = null;
  for (const v of arr) {
    const n = new Node(v);
    if (!head) head = n;
    if (prev) { prev.next = n; n.prev = prev; }
    prev = n;
  }
  return head;
}

function toArray(head) {
  const out = [];
  while (head) { out.push(head.data); head = head.next; }
  return out;
}

// quick checks
let h = buildDLL([3,4,5]);
h = reverse(h);
console.log(toArray(h)); // [5,4,3]

h = buildDLL([75,122,59,196]);
h = reverse(h);
console.log(toArray(h)); // [196,59,122,75]
